drud/ddev
# Static configuration of webserver and dbserver localhost ports, bind only to localhost, fixes #1491, fixes #941, #642 too

## The Problem/Issue/Bug:

fixes #1491, fixes #941

OP #1491: For security reasons, host binding of the dbserver should only be localhost (127.0.0.1) not 0.0.0.0 (the default, all interfaces), because this allows other people on the network to possibly probe for the database connection, depending on firewall configuration.

OP #941: Keep a stable port number for the localhost ephemeral ports so that people can use standard techniques to access them.

#642: We've long talked about keeping a catalog of metadata even for projects that aren't running. This PR introduces that.

## How this PR Solves The Problem:

* add `host_db_port` and `host_webserver_port`to config.yaml, use those values in the docker-compose.yaml.
* When a new host_db_port is written, add it to ~/.ddev/global_config.yaml so it's blacklisted from being used again for another project.
* Change the generated docker-compose.yaml to use explicit binding to 127.0.0.1 instead of the implicit 0.0.0.0

This is actually an amazing step forward for us. For the first time it provides a global database of projects, as requested in #642 . It's now trivial to add additional project information in the global_config.yaml.

## Manual Testing Instructions:

* Experiment with host_db_port and host_webserver port, both generated and explicitly configured.
* Verify that one project can't poach another project's ports.
* Verify that used ports show up in ~/.ddev/global_config.yaml
* Verify that db and web ports can only be accessed on the localhost interface, not via other interfaces on the machine. (You can use `ifconfig` on macOS and Linux to view other interfaces). You can test with `telnet localhost 33434` for port 33434, `telnet 172.28.99.33 33434` for a denied access to the same port on the 172.28.99.33 IP address (you must use the one that pertains to your system). Alternately, check from another machine on your network, but this requires the firewall to be turned off for correctness.
* Verify maintenance of ~/.ddev/global_config.yaml when projects are removed and rm --omit-data (which should completely remove the host)

## Automated Testing Overview:

- [x] Test for util.GetFreePort()
- [x] Test for access to used ports. (The existing router tests have some code that should help). 
- [x] Test to make sure that one project's ports can be reused deliberately or accidentally by another one.
- [x] Explicitly test to make sure localhost mysql and http port can be connected to correctly. This was only caught by accident on TestWebserverType (Docker Toolbox)

## Release/Deployment notes:

We'll probably want to tell people to do a `ddev config` on existing projects, which will add the new items in config.yaml.
